test/e2e/internal/utils/exec.go

```diff
@@ -0,0 +1,120 @@
+package utils
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gexec"
+)
+
+// ExecOpts is an option used to execute a command
+// TODO: to test system level context, need an /etc, /usr dir
+// Either executing tests using a docker, or mounting an overlayfs and chroot before launching the subprocess
+type ExecOpts struct {
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+	Env    map[string]string
+}
+
+// CommandOpts is an option used by CommandGroup to execute a batch of command
+type CommandOpts struct {
+	ExecOpts
+	Description string
+	Args        []string
+	ShouldFail  bool
+	// Checker is an user-provided function used to validate the result of a Command
+	Checker func(CommandOpts, *gexec.Session)
+}
+
+// CommandGroup contains a batch of e2e command to be executed
+type CommandGroup []CommandOpts
+
+func (g CommandGroup) WithAuth(username, password string) CommandGroup {
+	for i, c := range g {
+		g[i].ExecOpts = c.WithAuth(username, password)
+	}
+	return g
+}
+
+func (g CommandGroup) WithUserDir(dir string) CommandGroup {
+	for i, c := range g {
+		g[i].ExecOpts = c.WithUserDir(dir)
+	}
+	return g
+}
+
+func (opts ExecOpts) WithAuth(username, password string) ExecOpts {
+	if opts.Env == nil {
+		opts.Env = make(map[string]string)
+	}
+	opts.Env["NOTATION_USERNAME"] = username
+	opts.Env["NOTATION_PASSWORD"] = password
+	return opts
+}
+
+func (opts ExecOpts) WithUserDir(dir string) ExecOpts {
+	if opts.Env == nil {
+		opts.Env = make(map[string]string)
+	}
+	configDir, cacheDir := filepath.Join(dir, "config"), filepath.Join(dir, "cache")
+	os.MkdirAll(configDir, os.ModePerm)
+	os.MkdirAll(cacheDir, os.ModePerm)
+	opts.Env["XDG_CONFIG_HOME"] = configDir
+	opts.Env["XDG_CACHE_HOME"] = cacheDir
+	return opts
+}
+
+func Exec(binary string, opts ExecOpts, args ...string) (*gexec.Session, error) {
+	cmd := exec.Command(binary, args...)
+	cmd.Env = append(cmd.Env, os.Environ()...)
+	for key, val := range opts.Env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%v=%v", key, val))
+	}
+	if opts.Stdin != nil {
+		cmd.Stdin = opts.Stdin
+	}
+	session, err := gexec.Start(cmd, opts.Stdout, opts.Stderr)
+	if err != nil {
+		return nil, err
+	}
+	session = session.Wait("30s")
```

Comment:
It's just an empirical value. On my local machine with 8 cores (which means ginkgo cli will run 8 cases in parallel), it will timeout after 5-10s. In GitHub Actions, my pipeline can only use 2 cores, so I set the timeout value to be 3-4x longer only to make it pass the CI.